Document how to use Kueue for Deployments (#2698)
kubernetes-sigs/kueue: site/content/en/docs/tasks/run/deployment.md

---
title: "Run Deployment"
linkTitle: "Deployment"
date: 2024-07-25
weight: 6
description: >
  Run a Deployment as a Kueue-managed job.
---

> This page shows how to leverage Kueue's scheduling and resource management

Reviewer: I'd suggest providing a little more context.  For example,

Although Kueue does not yet support managing a Deployment as a single workload, it is still possible to leverage Kueue's scheduling and resource management capabilities for the individual Pods of the Deployment.

capabilities when running Deployments.

> This guide is for [batch users](/docs/tasks#batch-user) that have a basic understanding of Kueue. For more information, see [Kueue's overview](/docs/overview).

Reviewer 2: I would suggest to introduce "Serving user", with the description. As a description maybe we can have 
"A serving user runs serving workloads, for example, to expose a trained AI/ML model for inference"

## Before you begin

> 1. By default, the integration for `v1/pod` is not enabled.
>    Learn how to [install Kueue with a custom manager configuration](/docs/installation/#install-a-custom-configured-released-version)
>    and enable the `pod` integration.

Reviewer 2: ```suggestion
1. Learn how to [install Kueue with a custom manager configuration](/docs/installation/#install-a-custom-configured-released-version).
2. Enable the Pod integration as described in [Run Plain Pods](...)
```
I think this should be enough for now here.

   A configuration for Kueue with enabled pod integration would look like follows:
   ```yaml
   apiVersion: config.kueue.x-k8s.io/v1beta1
   kind: Configuration
   integrations:
     frameworks:
      - "pod"
     podOptions:
       # You can change namespaceSelector to define in which 
       # namespaces kueue will manage the pods.
       namespaceSelector:
         matchExpressions:
         - key: kubernetes.io/metadata.name
           operator: NotIn
           values: [ kube-system, kueue-system ]
       # Kueue uses podSelector to manage pods with particular 
       # labels. The default podSelector will match all the pods. 
       podSelector:
         matchExpressions:
         - key: kueue-job
           operator: In
           values: [ "true", "True", "yes" ]
   ```

> 2. Kueue will run webhooks for all created pods if the pod integration is enabled. The webhook namespaceSelector could be 

Reviewer 2: Similarly, this seems like a repetition, I would skip it.

Reviewer 2: Yeah, I understand the intention, but I think this is a bit too much detail copy-pasted, imo. 

Maybe in the introduction we could add some intro like:
```
In this section we demonstrate how to support scheduling Deployments in Kueue 
based on the Plain Pod integration, where every Pod from a Deployment is represented
as a single independent Plain Pod. This approach allows independent resource
management for the Pods, and thus scale up and down of the Deployment.
```
Then, in the "Before you begin" section something like: 
```
Follow steps in <link> to learn how to enable the ``v1/pod` integration and how to 
configure it using the `podOptions` field.
```

   used to filter the pods to reconcile. The default webhook namespaceSelector is:
   ```yaml
   matchExpressions:
   - key: kubernetes.io/metadata.name
     operator: NotIn
     values: [ kube-system, kueue-system ]
   ```
   
   When you [install Kueue via Helm](/docs/installation/#install-via-helm), the webhook namespace selector 
   will match the `integrations.podOptions.namespaceSelector` in the `values.yaml`.

   Make sure that namespaceSelector never matches the kueue namespace, otherwise the 
   Kueue deployment won't be able to create Pods.

> 3. Pods that belong to other API resources managed by Kueue are excluded from being queued by `pod` integration. 

Reviewer 2: I don't think this is relevant either.

   For example, pods managed by `batch/v1.Job` won't be managed by `pod` integration.

4. Check [Administer cluster quotas](/docs/tasks/manage/administer_cluster_quotas) for details on the initial Kueue setup.

> ## Running a Deployment admitted by Kueue

Reviewer 2: I would suggest to have a short sub-section of "Scale up or down", and mention: 

"You may perform scale up or scale down operations on Deployments. 
On scale down the exceess Pods are deleted and the quota is freed. 
On scale up new Pods are created and remain suspended until their corresponding workloads get admitted.
If there is not enough quota in your cluster, then the Deployment might be running only a subset of Pods."

When running Deployment on Kueue, take into consideration the following aspects:

### a. Queue selection

> The target [local queue](/docs/concepts/local_queue) should be specified in the `spec.template.metadata.labels` section of the Deployment configuration.

Reviewer: Maybe worth emphasizing that this is different than all the other "natively" supported kinds where you put the local queue in the metadata.labels.

```yaml
spec:
   template:
      metadata:
         labels:
            kueue.x-k8s.io/queue-name: user-queue
```

### b. Configure the resource needs

The resource needs of the workload can be configured in the `spec.template.spec.containers`.

```yaml
    - resources:
        requests:
          cpu: 3
```

### c. The "managed" label

> Kueue will inject the `kueue.x-k8s.io/managed=true` label to indicate which pods are managed by it.

Reviewer 2: I don't think this is important for users.

### d. Limitations

> - A Kueue managed Deployment cannot be created in `kube-system` or `kueue-system` namespaces.

Reviewer 2: I don't think this is an important limitation to mention here. Also, it can be altered by the `podOptions` so it is not really a strong limitation.

Reviewer: Maybe the limitation to mention is that the scope for Deployments  is implied by the pod integration's namespace selector.  No independent control for deployments.

> - In case of [preemption](/docs/concepts/cluster_queue/#preemption), the Deployment will

Reviewer 2: I don't think this is accurate - the Deployement would not be deleted, because all of its pods are independent units of preemption.

Reviewer: I agree with @mimowo -- I don't see how this would be the case since Kueue has no notion of the Deployment as a schedulable unit.

  be terminated and deleted.

## Example Deployment

Here is a sample Deployment:

{{< include "examples/deployment-kueue/kueue-deployment.yaml" "yaml" >}}

You can create the Deployment using the following command:
```sh

> # Create the pod

Reviewer 2: drop the comment

---
title: "Run Deployment"
linkTitle: "Deployment"
date: 2024-07-25
weight: 6
description: >
  Run a Deployment as a Kueue-managed workload.
---

This page shows how to leverage Kueue's scheduling and resource management
capabilities when running Deployments.
Although Kueue does not yet support managing a Deployment as a single Workload, 
it's still possible to leverage Kueue's scheduling and resource management capabilities for the individual Pods of the Deployment.

We demonstrate how to support scheduling Deployments in Kueue based on the Plain Pod integration,
where every Pod from a Deployment is represented as a single independent Plain Pod.

> This approach allows independent resource management for the Pods, and thus scale up and down of the Deployment.

Reviewer 2: ```suggestion
This approach allows independent resource management for the Pods, and thus scale-out and scale-in of the Deployment.
```

---
title: "Run Deployment"
linkTitle: "Deployment"
date: 2024-07-25
weight: 6
description: >

>   Run a Deployment as a Kueue-managed job.

Reviewer 3: ```suggestion
  Run a Deployment as a Kueue-managed workload.
```

---

This page shows how to leverage Kueue's scheduling and resource management
capabilities when running Deployments.

> Although Kueue does not yet support managing a Deployment as a single workload, 

Reviewer 3: ```suggestion
Although Kueue does not yet support managing a Deployment as a single Workload, 
```

I think that this `Workload` indicates the Kueue's Workload, right?

it's still possible to leverage Kueue's scheduling and resource management capabilities for the individual Pods of the Deployment.

> In this section we demonstrate how to support scheduling Deployments in Kueue based on the Plain Pod integration,

Reviewer 2: ```suggestion
We demonstrate how to support scheduling Deployments in Kueue based on the Plain Pod integration,
```
since we already mention above "This page shows" setting the context.

where every Pod from a Deployment is represented as a single independent Plain Pod.
This approach allows independent resource management for the Pods, and thus scale up and down of the Deployment.

This guide is for [serving users](/docs/tasks#serving-user) that have a basic understanding of Kueue.
For more information, see [Kueue's overview](/docs/overview).

## Before you begin

1. Learn how to [install Kueue with a custom manager configuration](/docs/installation/#install-a-custom-configured-released-version).

2. Follow steps in [Run Plain Pods](/docs/tasks/run/plain_pods/#before-you-begin)
to learn how to enable the `v1/pod` integration and how to configure it using the `podOptions` field.

> 3. Kueue will run webhooks for all created pods if the pod integration is enabled. The webhook namespaceSelector could be 

Reviewer 2: I think this can be dropped either. It is also part of the doc for Pod integration.

Reviewer 3: +1

   Make sure that namespaceSelector never matches the kueue namespace, otherwise the 
   Kueue deployment won't be able to create Pods.

4. Check [Administer cluster quotas](/docs/tasks/manage/administer_cluster_quotas) for details on the initial Kueue setup.

## Running a Deployment admitted by Kueue

When running Deployment on Kueue, take into consideration the following aspects:

### a. Queue selection

The target [local queue](/docs/concepts/local_queue) should be specified in the `spec.template.metadata.labels` section of the Deployment configuration. 
Since Kueue's scheduling and resource management will be applied to the individual Pods of the Deployment,
the queue name should be specified at the Pod level.

```yaml
spec:
   template:
      metadata:
         labels:
            kueue.x-k8s.io/queue-name: user-queue
```

### b. Configure the resource needs

The resource needs of the workload can be configured in the `spec.template.spec.containers`.

```yaml
    - resources:
        requests:
          cpu: 3
```

### c. Scaling

You may perform scale up or scale down operations on Deployments.

> On scale down the excess Pods are deleted and the quota is freed.

Reviewer 3: ```suggestion
On scale-in, the excess Pods are deleted, and the quota is freed.
```

The `scale down` indicates vertical, and the `scale in` indicates the horizontal.

Reviewer 2: Do you have some reference for that? I'm looking at the k8s docs https://kubernetes.io/docs/concepts/workloads/controllers/deployment/, and I uses of scale up / down.

Reviewer 3: Oh, I didn't know that.
But, in general, we use the scale-up for the vertical like this: https://azure.microsoft.com/en-us/resources/cloud-computing-dictionary/scaling-out-vs-scaling-up

Reviewer 2: I see, I'm ok with the proposed naming, it makes sense to discriminate.

> On scale up new Pods are created and remain suspended until their corresponding workloads get admitted.

Reviewer 3: ```suggestion
On scale-out, new Pods are created and remain suspended until their corresponding workloads get admitted.
```

The `scale up` indicates vertical, and the `scale out` indicates the horizontal.

> If there is not enough quota in your cluster, then the Deployment might be running only a subset of Pods.
> 

Reviewer 3: ```suggestion
If there is not enough quota in your cluster, the Deployment might run only a subset of Pods. So, if your workloads are business-critical, you can consider reserving the quota only for the serving workloads by the ClusterQueue `lendingLimit`. The `lendingLimit` allows you to rapidly scale out the critical serving workload.

For more `lendingLimit` details, please see the [ClusterQueue page](docs/concepts/cluster_queue#lendinglimit).
```

I would recommend using the `lendingLimit` for the critical serving workloads: https://kueue.sigs.k8s.io/docs/reference/kueue.v1beta1/#kueue-x-k8s-io-v1beta1-ResourceQuota

Reviewer 2: +1

### d. Limitations

- The scope for Deployments is implied by the pod integration's namespace selector. There's no independent control for deployments.

> ## Example Deployment

Reviewer 2: ```suggestion
## Example
```
